### PR TITLE
Additional `wlmon` fixes + `erase_stress_example`

### PR DESCRIPTION
**Pull request by omnitex**
**This branch contains up-to-date `wlmon`, version in`main` has issues**
**GUI is separate in [python-fe](/omnitex/espwlmon/pull/27) even though README mentions it here**

So the workflow with the currents PRs is this:
1. checkout this `wl-advanced-tweaks-and-fixes` branch
2. follow the README up to and including step 3
3. for step 4 switch to [python-fe](/omnitex/espwlmon/pull/27)

- readme changes will conflict with `python-fe` branch, let this be the newer version (but without `requirements.txt` here)
- [x] ~~the example might benefit from separate PR~~ nah
- [x] ~~move `wear-levelling` up a dir, get rid of `data-collector`? or keep for semantics?~~ keep

### CODE REVIEW

**Comment by omnitex** [comment]
- [ ] just now found out I might now be calculating `addr_erase_counts` to properly read it, somehow it previously worked without it???

**Comment by omnitex** [comment]
because all advanced read/write/erase operations use `calcAddr()` they also get randomized addresses. now that means reserved sectors (state, config etc.) might not actually be at the end of partition. and its a mess of which version of function is called when (advanced vs base vs some other?). detecting base WL by 0 Feistel keys is ok, calling `initSections()` does not initialize properly.

but erasing flash before flashing project with `WL_Advanced` turned on, works. so TODO investigate more

File: data-collector/wear_levelling/wlmon/main/include/wlmon.h [hunk 1]
```diff
@@ -12,9 +12,7 @@
 #define WLMON_DEFAULT_BUF_SIZE 500
 #define WLMON_BUF_SIZE 4096
 
-#define WL_MODE_UNDEFINED 0x0
-#define WL_MODE_BASE 0x1
-#define WL_MODE_ADVANCED 0x2
+typedef enum WL_MODE { WL_MODE_UNDEFINED = 0, WL_MODE_BASE, WL_MODE_ADVANCED } wl_mode_t;
```
Comment (pacucha42) [inline data-collector/wear_levelling/wlmon/main/include/wlmon.h:15]: Nice! Using enumerations is always preferred in such cases

File: erase_stress_example/main/erase_stress_example.c [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    const esp_vfs_fat_mount_config_t mount_config = {
+        .max_files = 4,
+        .format_if_mount_failed = true,
+        .allocation_unit_size = CONFIG_WL_SECTOR_SIZE
+    };
+
+    esp_err_t ret;
+    int full_runs = 0;
+
+start:
```
Comment (pacucha42) [inline erase_stress_example/main/erase_stress_example.c:41]: goto not really needed here - see my comments in the simulator code

File: erase_stress_example/main/erase_stress_example.c [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    storage = esp_partition_find_first(ESP_PARTITION_TYPE_DATA, ESP_PARTITION_SUBTYPE_ANY, "storage");
+    if (!storage) {
+        ESP_LOGE(TAG, "Failed to find first data partition labeled 'storage'");
+        return;
+    }
+
+    ESP_ERROR_CHECK(wl_mount(storage, &s_wl_handle));
+
+    // perform iterations of erasing a single sector to artificially stress the memory
+    #define ERASE_ITERATIONS (16*50)
```
Comment (pacucha42) [inline erase_stress_example/main/erase_stress_example.c:108]: #define perhaps not needed here - const definition instead?

File: erase_stress_example/main/erase_stress_example.c [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+            if (rand % 100 < 20) {
+                ESP_LOGI(TAG, "%s: will remount cycle at iteration %u", __func__, i);
+                for (int j = 0; j < (rand % 100 + 15); j++) {
+                    // unmounting always moves dummy sector
+                    ESP_ERROR_CHECK(wl_unmount(s_wl_handle));
+                    ESP_ERROR_CHECK(wl_mount(storage, &s_wl_handle));
+                }
+            }
+            // erase a sector around the middle of wear leveled partition
+            wl_erase_range(s_wl_handle, wl_size(s_wl_handle)/2 + (i%10 + rand%10), CONFIG_WL_SECTOR_SIZE);
```
Comment (pacucha42) [inline erase_stress_example/main/erase_stress_example.c:132]: missing retval check. general note for whole code (wherever applicable)

**Comment by pacucha42** [comment]
Still missing the default "showcase" application

File: erase_stress_example/main/erase_stress_example.c (hunk 2 above)
Comment (omnitex) [reply erase_stress_example/main/erase_stress_example.c]: `goto` is no more

File: erase_stress_example/main/erase_stress_example.c [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+            uint32_t rand = esp_random();
+
+            // random early break allows for some variance in erase counts
+            if (rand % 100 < 5) {
+                ESP_LOGI(TAG, "%s: random cycle break at iteration %u", __func__, i);
+                break;
+            }
+
+            // or remounting will move dummy sector 'a lot', making move_count increment (and in turn mapping shift) occur sooner
+            if (rand % 100 < 20) {
```
[the remark this answers is not in the document]
Comment (omnitex) [reply erase_stress_example/main/erase_stress_example.c]: tiny refactor + comments, should be clear now

File: erase_stress_example/main/erase_stress_example.c (hunk 4 above)
Comment (omnitex) [reply erase_stress_example/main/erase_stress_example.c]: yoinked a result check macro from WL which now checks wl erases/mounts/unmounts

**Pull request merged**